refactor(structure): make extract_f0_print functions to extract features

413x1nkp opened the pull request:
instead of just opening a subshell that calls the extract_f0_print.py, we use a dedicated function from the file to extract the f0

this improves security and speed quite significantly

Partially implements #49

413x1nkp committed:
refactor(structure): make extract_f0_print function to extract features

instead of just opening a subshell that calls the extract_f0_print.py, we use a dedicated function from the file to extract the f0

this improves security and speed quite significantly

413x1nkp committed:
refactor(structure): unite extract_f0_rmvpe and extract_f0_print into one

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp committed:
Merge pull request #7 from alexlnkp/formatter-dev

chore(format): run black on dev

413x1nkp committed:
refactor(structure): merge extract_f0_rmvpe_dml into extract_f0_print

now all of the f0 extractions are united in the extract_f0_print.py file

413x1nkp commented:
`printt` function's signature is changed, temporarily made it just print to stdout and not write to any logs.

Honestly, no idea how to make the `printt` work the way it used to. I thought of making lambdas and passing them as  arguments to initialization of the FeatureInput objects, but doesn't seem ideal. Ideas?

fumiama commented on infer/modules/train/extract/extract_f0_print.py:
```diff
@@ -16,19 +16,84 @@
-    f.write("%s\n" % strr)
-    f.flush()
 
+class DMLFeatureInputRMVPE(object):
```
As for the f0s, you can use the classes in rvc.f0. I have almost finished them. Meanwhile, be careful to compare the difference between rvc.f0 and code in the training.

413x1nkp replied:
> As for the f0s, you can use the classes in rvc.f0. I have almost finished them. Meanwhile, be careful to compare the difference between rvc.f0 and code in the training.

Speaking of which, shouldn't the `rvc/f0/__init__.py` have
```
__all__ = ["F0Predictor", "CRePE", "Dio", "Harvest", "PM", "RMVPE"]
```
In it?

fumiama replied:
Still WIP. You can open another PR.

fumiama commented on infer/modules/train/extract/extract_f0_print.py:
```diff
@@ -16,19 +16,84 @@
 
 def printt(strr):
     print(strr)
-    f.write("%s\n" % strr)
```
Well, print to stdout is also not some good practice. Maybe you can use the logging lib as the other codes did and pass a parameter to make it possible to disable the printing.
We also want to display the log at the web ui, so a custom callback may be needed.

413x1nkp replied:
Replaced with:
```py
logger = logging.getLogger("f0-feature-extract")

from multiprocessing import Process

# exp_dir = sys.argv[1]
# f = open("%s/extract_f0_feature.log" % exp_dir, "a+")


def printt(strr):
    logger.info(strr)
```
Looks good?

fumiama replied:
The name can be simplified. And you can search about a function hook in logger.

fumiama commented on web.py:
```diff
@@ -49,7 +55,7 @@
 os.environ["TEMP"] = tmp
 warnings.filterwarnings("ignore")
-torch.manual_seed(114514)
+torch.manual_seed(114514)  # ????
```
This is a meme number in the Chinese and Japanese Internet.

413x1nkp replied:
> This is a meme number in the Chinese and Japanese Internet.

That makes a lot more sense to be honest lol

fumiama replied:
And also a number 23333, if you see some other codes.

413x1nkp committed:
fix: make rmvpe_gpu valid f0 method

fumiama reviewed (commented):
Just a glance. Good job but you may find it painful to follow my modification on the rvc lib 😂. You can check the rvc folder and see what functions/class can be reused.

413x1nkp committed:
refactor: remove bunch of repeating code

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp committed:
Merge pull request #8 from alexlnkp/formatter-dev

chore(format): run black on dev

413x1nkp committed:
refactor: secret magic number

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
refactor: use classes in rvc.f0 for f0 computation

413x1nkp committed:
refactor: extract_features code refactoring

413x1nkp commented:
Some of the functions in f0 submodule use type subscription that isn't available in python 3.8, hence the checks are unsuccessful.

Other than that, i've stumbled upon a really odd problem, `isinstance` does not seem to be working for some reason?
```py
    if isinstance(predictor, RMVPE):
        predictor = predictor("assets/rmvpe/rmvpe.pt", is_half, device)
    else:
        predictor = predictor(device=device)
    featureInput = predictor
```
And it throws:
```
Traceback (most recent call last):
  File "/usr/lib/python3.12/threading.py", line 1073, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.12/threading.py", line 1010, in run
    self._target(*self._args, **self._kwargs)
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/train/extract/extract_f0_print.py", line 108, in call_extract_features
    extract_features(predictors[method_f0], expected_dir, is_half, device)
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/modules/train/extract/extract_f0_print.py", line 53, in extract_features
    predictor = predictor(device=device)
                ^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: RMVPE.__init__() missing 2 required positional arguments: 'model_path' and 'is_half'
```
Even though it should've been initialized with `predictor = predictor("assets/rmvpe/rmvpe.pt", is_half, device)`, since it would be an instance of `RMVPE`

EDIT: Using explicit type checking (i.e. `type(B) == type(A)` or `B is A`) works. Turns out it's an issue because RMVPE derives from F0Predictor :)

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
fix: use explicit type check instead of `isinstance()`

due to inheritance `isinstance(predictor, RMVPE)` is always false, since it's also derived from F0Predictor

fumiama commented:
> Some of the functions in f0 submodule use type subscription that isn't available in python 3.8

You can open a PR to fix it since I'm using 3.9 😂

> isinstance does not seem to be working for some reason

That is a class name but not an instance. Try just use `is`.

413x1nkp committed:
refactor: use `is` instead of `type(B) == type(A)` for cleaner code

413x1nkp commented:
> > Some of the functions in f0 submodule use type subscription that isn't available in python 3.8
> 
> You can open a PR to fix it since I'm using 3.9 😂
> 
> > isinstance does not seem to be working for some reason
> 
> That is a class name but not an instance. Try just use `is`.

I'll make a PR for the type subscription error fix on 3.8

Yeah, that was the problem, lol. All of this polymorphism stuff is a bit confusing sometimes

413x1nkp commented:
Now we're getting somewhere!
```
2024-06-14 18:45:00 | INFO | rvc.F0Print | Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/10_46.wav: 'str' object has no attribute 'shape'
2024-06-14 18:45:00 | INFO | rvc.F0Print | Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/10_48.wav: 'str' object has no attribute 'shape'
2024-06-14 18:45:00 | INFO | rvc.F0Print | Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/10_5.wav: 'str' object has no attribute 'shape'
2024-06-14 18:45:00 | INFO | rvc.F0Print | Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/10_50.wav: 'str' object has no attribute 'shape'
2024-06-14 18:45:00 | INFO | rvc.F0Print | Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/10_52.wav: 'str' object has no attribute 'shape'
2024-06-14 18:45:00 | INFO | rvc.F0Print | Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/10_6.wav: 'str' object has no attribute 'shape'
2024-06-14 18:45:00 | INFO | rvc.F0Print | Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/10_7.wav: 'str' object has no attribute 'shape'

. . .
```

Time to find out why this is happening and the f0-feature-print is up and running again!

413x1nkp committed:
refactor: rename variables and functions to be less confusing

413x1nkp committed:
fix: use `load_audio` function to get numpy array representing the audio at the path provided

413x1nkp commented:
This is rather strange
```
2024-06-14 18:54:48 | INFO | rvc.F0Print | Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/10_6.wav: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method
```
Replacing the thread creation with:
```py
    for idx, (inp_path, coarse_path, feature_path) in enumerate(paths):
        p = mp.spawn(fn=save_f0, args=(featureInput, inp_path, coarse_path, feature_path), join=False)
        ps.append(p)
```
Leads to... peculiar behavior
```
2024-06-14 19:03:26 | INFO | configs.config | Found GPU NVIDIA GeForce RTX 3060
2024-06-14 19:03:26 | INFO | configs.config | Half-precision floating-point: True, device: cuda:0
2024-06-14 19:03:27 | INFO | infer.lib.rvcmd | checking hubret & rmvpe...
2024-06-14 19:03:27 | INFO | infer.lib.rvcmd | checking pretrained models...
2024-06-14 19:03:28 | INFO | infer.lib.rvcmd | checking pretrained models v2...
2024-06-14 19:03:29 | INFO | infer.lib.rvcmd | checking uvr5_weights...
2024-06-14 19:03:29 | INFO | infer.lib.rvcmd | all assets are already latest.
2024-06-14 19:03:29 | INFO | __main__ | Language: en_US
2024-06-14 19:03:40 | INFO | configs.config | Found GPU NVIDIA GeForce RTX 3060
2024-06-14 19:03:40 | INFO | configs.config | Half-precision floating-point: True, device: cuda:0
2024-06-14 19:03:41 | INFO | infer.lib.rvcmd | checking hubret & rmvpe...
2024-06-14 19:03:42 | INFO | infer.lib.rvcmd | checking pretrained models...
2024-06-14 19:03:42 | INFO | infer.lib.rvcmd | checking pretrained models v2...
2024-06-14 19:03:43 | INFO | infer.lib.rvcmd | checking uvr5_weights...
2024-06-14 19:03:44 | INFO | infer.lib.rvcmd | all assets are already latest.
2024-06-14 19:03:44 | INFO | __mp_main__ | Language: en_US
2024-06-14 19:03:44 | ERROR | __mp_main__ | Cannot find empty port in range: 7865-7865. You can specify a different port by setting the GRADIO_SERVER_PORT environment variable or passing the `server_port` parameter to `launch()`.
2024-06-14 19:03:46 | INFO | configs.config | Found GPU NVIDIA GeForce RTX 3060
2024-06-14 19:03:46 | INFO | configs.config | Half-precision floating-point: True, device: cuda:0
2024-06-14 19:03:47 | INFO | infer.lib.rvcmd | checking hubret & rmvpe...
2024-06-14 19:03:47 | INFO | infer.lib.rvcmd | checking pretrained models...
2024-06-14 19:03:48 | INFO | infer.lib.rvcmd | checking pretrained models v2...
2024-06-14 19:03:49 | INFO | infer.lib.rvcmd | checking uvr5_weights...
2024-06-14 19:03:49 | INFO | infer.lib.rvcmd | all assets are already latest.
2024-06-14 19:03:49 | INFO | __mp_main__ | Language: en_US
2024-06-14 19:03:50 | ERROR | __mp_main__ | Cannot find empty port in range: 7865-7865. You can specify a different port by setting the GRADIO_SERVER_PORT environment variable or passing the `server_port` parameter to `launch()`.
2024-06-14 19:03:52 | INFO | configs.config | Found GPU NVIDIA GeForce RTX 3060
2024-06-14 19:03:52 | INFO | configs.config | Half-precision floating-point: True, device: cuda:0
2024-06-14 19:03:53 | INFO | infer.lib.rvcmd | checking hubret & rmvpe...
2024-06-14 19:03:53 | INFO | infer.lib.rvcmd | checking pretrained models...
2024-06-14 19:03:54 | INFO | infer.lib.rvcmd | checking pretrained models v2...
2024-06-14 19:03:55 | INFO | infer.lib.rvcmd | checking uvr5_weights...
2024-06-14 19:03:55 | INFO | infer.lib.rvcmd | all assets are already latest.
2024-06-14 19:03:55 | INFO | __mp_main__ | Language: en_US
2024-06-14 19:03:55 | ERROR | __mp_main__ | Cannot find empty port in range: 7865-7865. You can specify a different port by setting the GRADIO_SERVER_PORT environment variable or passing the `server_port` parameter to `launch()`.
2024-06-14 19:03:58 | INFO | configs.config | Found GPU NVIDIA GeForce RTX 3060
2024-06-14 19:03:58 | INFO | configs.config | Half-precision floating-point: True, device: cuda:0
2024-06-14 19:03:59 | INFO | infer.lib.rvcmd | checking hubret & rmvpe...
2024-06-14 19:03:59 | INFO | infer.lib.rvcmd | checking pretrained models...
2024-06-14 19:04:00 | INFO | infer.lib.rvcmd | checking pretrained models v2...
2024-06-14 19:04:01 | INFO | infer.lib.rvcmd | checking uvr5_weights...
2024-06-14 19:04:01 | INFO | infer.lib.rvcmd | all assets are already latest.
2024-06-14 19:04:01 | INFO | __mp_main__ | Language: en_US
2024-06-14 19:04:01 | ERROR | __mp_main__ | Cannot find empty port in range: 7865-7865. You can specify a different port by setting the GRADIO_SERVER_PORT environment variable or passing the `server_port` parameter to `launch()`.
```

EDIT: added these lines and removed the basic `multiprocessing` import:
```py
from torch.multiprocessing import Process, set_start_method
set_start_method('spawn', force=True)
```
Process creation looks like this now:
```py
    ps = []
    for idx, (inp_path, coarse_path, feature_path) in enumerate(paths):
        p = Process(name=f"extract_f0_feature_{idx}", target=save_f0, args=(featureInput, inp_path, coarse_path, feature_path))
        p.start()
        ps.append(p)

    for p in ps:
        p.join()
```

So, this should work, but instead it still does the same odd thing
```
# After starting f0 extraction
2024-06-14 19:03:40 | INFO | configs.config | Found GPU NVIDIA GeForce RTX 3060
2024-06-14 19:03:40 | INFO | configs.config | Half-precision floating-point: True, device: cuda:0
2024-06-14 19:03:41 | INFO | infer.lib.rvcmd | checking hubret & rmvpe...
2024-06-14 19:03:42 | INFO | infer.lib.rvcmd | checking pretrained models...
2024-06-14 19:03:42 | INFO | infer.lib.rvcmd | checking pretrained models v2...
2024-06-14 19:03:43 | INFO | infer.lib.rvcmd | checking uvr5_weights...
2024-06-14 19:03:44 | INFO | infer.lib.rvcmd | all assets are already latest.
2024-06-14 19:03:44 | INFO | __mp_main__ | Language: en_US
2024-06-14 19:03:44 | ERROR | __mp_main__ | Cannot find empty port in range: 7865-7865. You can specify a different port by setting the GRADIO_SERVER_PORT environment variable or passing the `server_port` parameter to `launch()`.
2024-06-14 19:03:40 | INFO | configs.config | Found GPU NVIDIA GeForce RTX 3060
2024-06-14 19:03:40 | INFO | configs.config | Half-precision floating-point: True, device: cuda:0
2024-06-14 19:03:41 | INFO | infer.lib.rvcmd | checking hubret & rmvpe...
2024-06-14 19:03:42 | INFO | infer.lib.rvcmd | checking pretrained models...
2024-06-14 19:03:42 | INFO | infer.lib.rvcmd | checking pretrained models v2...
2024-06-14 19:03:43 | INFO | infer.lib.rvcmd | checking uvr5_weights...
2024-06-14 19:03:44 | INFO | infer.lib.rvcmd | all assets are already latest.
2024-06-14 19:03:44 | INFO | __mp_main__ | Language: en_US
2024-06-14 19:03:44 | ERROR | __mp_main__ | Cannot find empty port in range: 7865-7865. You can specify a different port by setting the GRADIO_SERVER_PORT environment variable or passing the `server_port` parameter to `launch()`.
```

I guess it decides to duplicate the WHOLE thread? Not sure what to do about that, to be honest
I guess we should add `if __name__ == "__main__"` to the web.py?

413x1nkp committed:
Merge branch 'dev' into dev

fumiama commented:
> I guess we should add if __name__ == "__main__" to the web.py

You can test it. But instead of
```py
if __name__ == "__main__":
   ...
```
you can use
```py
if __name__ != "__main__":
    return
```
At the very beginning before all imports.

fumiama commented:
Also, I have added fcpe to the f0 so maybe you can try to let this code support for all f0s available.

413x1nkp committed:
fix: fix f0 predictors

all are working except for rmvpe

413x1nkp committed:
Merge branch 'dev' into dev

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp commented:
> Also, I have added fcpe to the f0 so maybe you can try to let this code support for all f0s available.

Added `fcpe` as a valid f0 extractor in 814a792 :)

> You can test it. But instead of
> 
> ```python
> if __name__ == "__main__":
>    ...
> ```
> 
> you can use
> 
> ```python
> if __name__ != "__main__":
>     return
> ```
> 
> At the very beginning before all imports.

This works, but not for RMVPE or RMVPE_GPU. I assume the issue is not with how those two are called, but rather the way they predict the f0.

Dio, PM and Harvest are working as expected. (note the `logger.info(f"Starting extract_f0_feature_{idx} thread for {inp_path}")`, I'll remove it later or will replace with `logger.debug(f"Starting extract_f0_feature_{idx} thread for {inp_path}")`)

RMVPE is failing to create even a single thread for some reason...

413x1nkp committed:
Merge pull request #9 from alexlnkp/formatter-dev

chore(format): run black on dev

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
fix: fix f0 extraction logging

provide a few docstrings for the functions because i suck at naming things properly

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp committed:
Merge pull request #10 from alexlnkp/formatter-dev

chore(format): run black on dev

413x1nkp commented:
I think it's done? The RMVPE still doesn't want to extract f0 features at all, but I'm not sure there's something i can do about that in particular, I'm not very knowledgeable in the inner workings of f0 extractors in general...

There are no error messages though, however the thread is just failing to join, I assume...

Please take a look at this and let me know what you think

fumiama commented on infer/modules/train/extract/extract_f0_print.py:
```diff
@@ -1,175 +1,151 @@
+    elif predictor is CRePE or predictor is FCPE:
+        predictor = predictor(device=device)
+    else:
+        predictor = predictor()
```
The parameter is important and you cannot use the default one. You can get these parameters from the original code.
ex. hop_len = 160, sr = 16000, etc.

413x1nkp replied:
fixed in 5d75a21

fumiama commented on web.py:
```diff
@@ -258,106 +264,36 @@ def preprocess_dataset(trainset_dir, exp_dir, sr, n_p):
+        while extracting_thread.is_alive():
+            yield getlog(log_path)
+
+        log = getlog(log_path)
```
I think it's not so good. Maybe use pipe/function callback to print the log.

413x1nkp replied:
I couldn't figure out exactly how you suggested me to do it so i made it extremely hacky :sob: 
```py
        extracting_thread.start()
        with open(log_path, "r") as f:
            while extracting_thread.is_alive():
                lines = f.readlines()
                for line in lines:
                    logger.info(line.strip("\n"))
                
                yield from lines

            yield f.read()
```
Commited in ca20b03, but attempting my hardest to improve the way it is...

fumiama replied:
I mean not read anything from the log file. That file is just a backup in case the program stopped suddenly and all logs disappeared. You should get the log string from another place ex. pipe or func calls.

413x1nkp replied:
> I mean not read anything from the log file. That file is just a backup in case the program stopped suddenly and all logs disappeared. You should get the log string from another place ex. pipe or func calls.

So, no reading from the log file at all during execution, just use regular logger instance to print info
understood

413x1nkp replied:
Fixed in 54b2d8d

fumiama reviewed (commented):
- I think it'd be better to compare the features extracted between old and your modification to ensure that they are the same.
- As for the RMVPE, we can find it out later.
- I suggest that you can implement the single-thread version first with function call only, then start to make it be multi-thread.

413x1nkp committed:
chore: pass the default parameters to the predictor in the extract_features

413x1nkp committed:
fix: clear log file before the f0 extraction

this is done so the previous logs are deleted and the new ones are yielded correctly for gradio to display

413x1nkp committed:
revert: make the logfile erased in the web.py instead of extract_f0_print.py

this should be better structurally

413x1nkp committed:
refactor: bad code, fix ASAP. log and yield contents of logfile conventionally

this should technically be more "correct" than yield "".join(lines), also tried to use logger in a better way

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp committed:
Merge pull request #11 from alexlnkp/formatter-dev

chore(format): run black on dev

fumiama commented on infer/modules/train/extract/extract_f0_print.py:
```diff
@@ -1,175 +1,155 @@
+        predictor = predictor("assets/rmvpe/rmvpe.pt", is_half, device)
+    elif predictor is CRePE or predictor is FCPE:
+        predictor = predictor(
+            hop_length=512, f0_min=50, f0_max=1100, sampling_rate=44100, device=device
```
Are you sure the sampling rate is 44100?

413x1nkp replied:
Whoops, sorry
You're right, it's 16000

fumiama replied:
Oh yes, and for the RMVPE, no need to specify those values. Just is_half and device is OK.

413x1nkp replied:
Fixed in 54b2d8d

fumiama commented on infer/modules/train/extract/extract_f0_print.py:
```diff
@@ -1,175 +1,155 @@
-            f0 = (
-                parselmouth.Sound(x, self.fs)
-                .to_pitch_ac(
-                    time_step=time_step / 1000,
```
See here for the reference.

fumiama commented on infer/modules/train/extract/extract_f0_print.py:
```diff
@@ -1,175 +1,155 @@
-        elif f0_method == "harvest":
-            f0, t = pyworld.harvest(
-                x.astype(np.double),
-                fs=self.fs,
```
You should not specify these values explicitly but use self.hop, self.f0_min, self.f0_max, etc.

413x1nkp replied:
Shouldn't those be used automatically if no value is passed?
When i didn't pass those values they were set to the same values they were defined with in the function, so i'm not really sure what exactly seems to be the problem...

fumiama replied:
> Shouldn't those be used automatically if no value is passed?

The automatic value is different from what currently used.

413x1nkp replied:
> > Shouldn't those be used automatically if no value is passed?
> 
> The automatic value is different from what currently used.

So... what's the problem again? ._.
I'm sorry I'm just confused..

fumiama replied:
You can see that in this original code, it used self.f0_max, self.f0_min, self.hop, etc. So you should also use that reference, not the absolute value.

413x1nkp replied:
> You can see that in this original code, it used self.f0_max, self.f0_min, self.hop, etc. So you should also use that reference, not the absolute value.

Do you mean something like this?
```py
        predictor = predictor(
            hop_length=predictor.hop_length, f0_min=predictor.f0_min, f0_max=predictor.f0_max, sampling_rate=16000, device=device
        )
```
Because i don't have a class FeatureInput like in the original code, I have no `self` of reference to pass as the value

I'm not really sure why even pass parameters if they're going to be the default ones...

fumiama replied:
OK. Well, the typical value should come from config.json but since the original code just specify it, you can specify it too. I will unify them later. Also, the hop_len is ALWAYS 160, don't use the value 512.

413x1nkp replied:
> OK. Well, the typical value should come from config.json but since the original code just specify it, you can specify it too. I will unify them later. Also, the hop_len is ALWAYS 160, don't use the value 512.

Yeah, I got confused because your new implementation of F0Predictor uses 512 by default
Changed to 160 in 2e405c2 though!

fumiama replied:
> I got confused because your new implementation of F0Predictor uses 512 by default

It is deliberately implemented to make sure all the parameters are filled in correctly. If not, the result will be incorrect.

413x1nkp committed:
fix: use 16khz samplerate for predictors instead of 44.1khz

this is the way those valued were originally

413x1nkp committed:
fix: don't read out the log into the stdout

413x1nkp committed:
feat(logging): improve logging

attempting to make the logging the way it was before this branch.

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
fix: hotfix of a typo

github-actions[bot] committed:
chore(format): run black on dev

413x1nkp committed:
Merge pull request #12 from alexlnkp/formatter-dev

chore(format): run black on dev

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
fix: use hop_length of 160 instead of 512

this is the way it was in the original code

fumiama commented on infer/modules/train/extract/extract_f0_print.py:
```diff
@@ -1,175 +1,170 @@
-import numpy as np
-import pyworld
+logging.getLogger("numba").setLevel(logging.WARNING)
+logger = logging.getLogger("rvc.F0Print")
```
You can use
```py
logging.getLogger(__name__)
```

413x1nkp replied:
Implemented that! :)

fumiama commented on infer/modules/train/extract/extract_f0_print.py:
```diff
@@ -1,175 +1,170 @@
+        __logfile = open(logfile, "w")
+
+        out_features = predictor.compute_f0(load_audio(inp_path, 16000))
+        out_coarse = 0  # TODO: add coarse f0
```
Have you done this?

413x1nkp replied:
> Have you done this?

Technically it works, yeah
However I'm having troubles saving features now, for some reason...

Working on it!

413x1nkp replied:
Now that features are saving properly, the same coarse function is being called on the `out_features`
707536e

fumiama commented on web.py:
```diff
@@ -1,5 +1,10 @@
+# A hack, fix later
```
You can use the TODO grammar to highlight it.

413x1nkp replied:
Did that in ba06cb6 :)

fumiama commented on web.py:
```diff
@@ -260,106 +267,36 @@ def preprocess_dataset(trainset_dir, exp_dir, sr, n_p):
+    open(log_path, "w").close()
+
     if if_f0:
-        if f0method != "rmvpe_gpu":
```
If it is OK to remove the gpu comparison, just remove the whole rmvpe_gpu and merge it into rmvpe.

413x1nkp replied:
> If it is OK to remove the gpu comparison, just remove the whole rmvpe_gpu and merge it into rmvpe.

To dynamically just use GPU if it's available? Sounds like a good idea to me.
Will do that after i fix features not saving

413x1nkp replied:
Did that in 0b95532

fumiama reviewed (commented):
- Are RMVPE and RMVPE_GPU usable?
- Have you compared the result of the refactored code to the original one and make sure they are the same?
- As for the log, can it be printed on the webui?

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp committed:
feat: add f0 coarse extraction

doesn't save .npy files properly yet.

413x1nkp commented:
> * Have you compared the result of the refactored code to the original one and make sure they are the same?

Doing that now, found out a lot of bugs...
```
5_48.wav/5_49.wav/5_5.wav/5_50.wav/5_51.wav/5_52.wav/5_53.wav/5_54.wav/5_55.wav/5_56.wav/5_57.wav/5_58.wav/5_6.wav/5_60.wav/5_7.wav/5_8.wav/5_9.wav/6_0.wav/6_1.wav/6_10.wav/6_11.wav/6_12.wav/6_13.wav/6_14.wav/6_15.wav/6_16.wav/6_17.wav/6_18.wav/6_19.wav/6_2.wav/6_20.wav/6_21.wav/6_22.wav/6_23.wav/6_24.wav/6_25.wav/6_26.wav/6_27.wav/6_28.wav/6_29.wav/6_3.wav/6_30.wav/6_31.wav/6_32.wav/6_33.wav/6_34.wav/6_35.wav/6_36.wav/6_37.wav/6_38.wav/6_39.wav/6_4.wav/6_40.wav/6_41.wav/6_42.wav/6_43.wav/6_44.wav/6_45.wav/6_47.wav/6_5.wav/6_6.wav/6_7.wav/6_8.wav/6_9.wav/7_0.wav/7_1.wav/7_10.wav/7_100.wav/7_102.wav/7_103.wav/7_104.wav/7_105.wav/7_106.wav/7_108.wav/7_11.wav/7_12.wav/7_13.wav/7_14.wav/7_15.wav/7_16.wav/7_17.wav/7_18.wav/7_19.wav/7_20.wav/7_21.wav/7_22.wav/7_23.wav/7_24.wav/7_25.wav/7_26.wav/7_27.wav/7_28.wav/7_29.wav/7_3.wav/7_30.wav/7_31.wav/7_32.wav/7_33.wav/7_34.wav/7_35.wav/7_36.wav/7_37.wav/7_38.wav/7_39.wav/7_40.wav/7_41.wav/7_43.wav/7_44.wav/7_45.wav/7_46.wav/7_47.wav/7_48.wav/7_49.wav/7_5.wav/7_50.wav/7_51.wav/7_52.wav/7_53.wav/7_54.wav/7_55.wav/7_56.wav/7_57.wav/7_58.wav/7_59.wav/7_6.wav/7_60.wav/7_61.wav/7_62.wav/7_63.wav/7_64.wav/7_65.wav/7_66.wav/7_67.wav/7_68.wav/7_69.wav
2024-06-17 20:52:08 | INFO | infer.modules.train.extract.extract_f0_print | /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/2b-f0nsf/0_0.wav/0_1.wav/0_10.wav/0_11.wav/0_12.wav/0_13.wav/0_14.wav/0_15.wav/0_16.wav/0_17.wav/0_18.wav/0_19.wav/0_2.wav/0_20.wav/0_21.wav/0_22.wav/0_23.wav/0_24.wav/0_25.wav/0_26.wav/0_27.wav/0_28.wav/0_29.wav/0_3.wav/0_30.wav/0_31.wav/0_32.wav/0_33.wav/0_35.wav/0_4.wav/0_5.wav/0_6.wav/0_7.wav/0_8.wav/0_9.wav/10_0.wav/10_1.wav/10_10.wav/10_11.wav/10_12.wav/10_13.wav/10_14.wav/10_15.wav/10_16.wav/10_17.wav/10_18.wav/10_2.wav/10_20.wav/10_21.wav/10_22.wav/10_23.wav/10_24.wav/10_25.wav/10_26.wav/10_27.wav/10_28.wav/10_29.wav/10_3.wav/10_30.wav/10_31.wav/10_32.wav/10_33.wav/10_34.wav/10_35.wav/10_36.wav/10_37.wav/10_38.wav/10_39.wav/10_4.wav/10_40.wav/10_41.wav/10_42.wav/10_43.wav/10_45.wav/10_46.wav/10_48.wav/10_5.wav/10_50.wav/10_52.wav/10_6.wav/10_7.wav/10_8.wav/10_9.wav/1_0.wav/1_1.wav/1_10.wav/1_11.wav/1_12.wav/1_13.wav/1_14.wav/1_15.wav/1_16.wav/1_17.wav/1_18.wav/1_19.wav/1_2.wav/1_20.wav/1_21.wav/1_22.wav/1_23.wav/1_24.wav/1_25.wav/1_26.wav/1_27.wav/1_28.wav/1_29.wav/1_3.wav/1_30.wav/1_31.wav/1_32.wav/1_33.wav/1_35.wav/1_4.wav/1_5.wav/1_6.wav/1_7.wav/1_8.wav/1_9.wav/2_0.wav/2_1.wav/2_10.wav/2_11.wav/2_12.wav/2_13.wav/2_14.wav/2_15.wav/2_16.wav/2_17.wav/2_18.wav/2_2.wav/2_3.wav/2_4.wav/2_5.wav/2_6.wav/2_7.wav/2_8.wav/2_9.wav/3_0.wav/3_1.wav/3_10.wav/3_11.wav/3_12.wav/3_13.wav/3_14.wav/3_15.wav/3_16.wav/3_18.wav/3_2.wav/3_3.wav/3_4.wav/3_5.wav/3_6.wav/3_7.wav/3_8.wav/3_9.wav/4_0.wav/4_10.wav/4_11.wav/4_13.wav/4_2.wav/4_3.wav/4_4.wav/4_5.wav/4_6.wav/4_7.wav/4_8.wav/4_9.wav/5_0.wav/5_1.wav/5_10.wav/5_11.wav/5_12.wav/5_13.wav/5_14.wav/5_15.wav/5_16.wav/5_17.wav/5_18.wav/5_19.wav/5_2.wav/5_20.wav/5_21.wav/5_22.wav/5_23.wav/5_24.wav/5_25.wav/5_26.wav/5_27.wav/5_28.wav/5_29.wav/5_3.wav/5_30.wav/5_31.wav/5_32.wav/5_33.wav/5_34.wav/5_35.wav/5_36.wav/5_37.wav/5_38.wav/5_39.wav/5_4.wav/5_40.wav/5_41.wav/5_42.wav/5_43.wav/5_44.wav/5_45.wav/5_46.wav/5_47.wav/5_48.wav/5_49.wav/5_5.wav/5_50.wav/5_51.wav/5_52.wav/5_53.wav/5_54.wav/5_55.wav/5_56.wav/5_57.wav/5_58.wav/5_6.wav/5_60.wav/5_7.wav/5_8.wav/5_9.wav/6_0.wav/6_1.wav/6_10.wav/6_11.wav/6_12.wav/6_13.wav/6_14.wav/6_15.wav/6_16.wav/6_17.wav/6_18.wav/6_19.wav/6_2.wav/6_20.wav/6_21.wav/6_22.wav/6_23.wav/6_24.wav/6_25.wav/6_26.wav/6_27.wav/6_28.wav/6_29.wav/6_3.wav/6_30.wav/6_31.wav/6_32.wav/6_33.wav/6_34.wav/6_35.wav/6_36.wav/6_37.wav/6_38.wav/6_39.wav/6_4.wav/6_40.wav/6_41.wav/6_42.wav/6_43.wav/6_44.wav/6_45.wav/6_47.wav/6_5.wav/6_6.wav/6_7.wav/6_8.wav/6_9.wav/7_0.wav/7_1.wav/7_10.wav/7_100.wav/7_102.wav/7_103.wav/7_104.wav/7_105.wav/7_106.wav/7_108.wav/7_11.wav/7_12.wav/7_13.wav/7_14.wav/7_15.wav/7_16.wav/7_17.wav/7_18.wav/7_19.wav/7_20.wav/7_21.wav/7_22.wav/7_23.wav/7_24.wav/7_25.wav/7_26.wav/7_27.wav/7_28.wav/7_29.wav/7_3.wav/7_30.wav/7_31.wav/7_32.wav/7_33.wav/7_34.wav/7_35.wav/7_36.wav/7_37.wav/7_38.wav/7_39.wav/7_40.wav/7_41.wav/7_43.wav/7_44.wav/7_45.wav/7_46.wav/7_47.wav/7_48.wav/7_49.wav/7_5.wav/7_50.wav/7_51.wav/7_52.wav/7_53.wav/7_54.wav/7_55.wav/7_56.wav/7_57.wav/7_58.wav/7_59.wav/7_6.wav/7_60.wav/7_61.wav/7_62.wav/7_63.wav/7_64.wav/7_65.wav/7_66.wav/7_67.wav/7_68.wav/7_69.wav/7_7.wav
2024-06-17 20:52:08 | INFO | infer.modules.train.extract.extract_f0_print | /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/2b-f0nsf/0_0.wav/0_1.wav/0_10.wav/0_11.wav/0_12.wav/0_13.wav/0_14.wav/0_15.wav/0_16.wav/0_17.wav/0_18.wav/0_19.wav/0_2.wav/0_20.wav/0_21.wav/0_22.wav/0_23.wav/0_24.wav/0_25.wav/0_26.wav/0_27.wav/0_28.wav/0_29.wav/0_3.wav/0_30.wav/0_31.wav/0_32.wav/0_33.wav/0_35.wav/0_4.wav/0_5.wav/0_6.wav/0_7.wav/0_8.wav/0_9.wav/10_0.wav/10_1.wav/10_10.wav/10_11.wav/10_12.wav/10_13.wav/10_14.wav/10_15.wav/10_16.wav/10_17.wav/10_18.wav/10_2.wav/10_20.wav/10_21.wav/10_22.wav/10_23.wav/10_24.wav/10_25.wav/10_26.wav/10_27.wav/10_28.wav/10_29.wav/10_3.wav/10_30.wav/10_31.wav/10_32.wav/10_33.wav/10_34.wav/10_35.wav/10_36.wav/10_37.wav/10_38.wav/10_39.wav/10_4.wav/10_40.wav/10_41.wav/10_42.wav/10_43.wav/10_45.wav/10_46.wav/10_48.wav/10_5.wav/10_50.wav/10_52.wav/10_6.wav/10_7.wav/10_8.wav/10_9.wav/1_0.wav/1_1.wav/1_10.wav/1_11.wav/1_12.wav/1_13.wav/1_14.wav/1_15.wav/1_16.wav/1_17.wav/1_18.wav/1_19.wav/1_2.wav/1_20.wav/1_21.wav/1_22.wav/1_23.wav/1_24.wav/1_25.wav/1_26.wav/1_27.wav/1_28.wav/1_29.wav/1_3.wav/1_30.wav/1_31.wav/1_32.wav/1_33.wav/1_35.wav/1_4.wav/1_5.wav/1_6.wav/1_7.wav/1_8.wav/1_9.wav/2_0.wav/2_1.wav/2_10.wav/2_11.wav/2_12.wav/2_13.wav/2_14.wav/2_15.wav/2_16.wav/2_17.wav/2_18.wav/2_2.wav/2_3.wav/2_4.wav/2_5.wav/2_6.wav/2_7.wav/2_8.wav/2_9.wav/3_0.wav/3_1.wav/3_10.wav/3_11.wav/3_12.wav/3_13.wav/3_14.wav/3_15.wav/3_16.wav/3_18.wav/3_2.wav/3_3.wav/3_4.wav/3_5.wav/3_6.wav/3_7.wav/3_8.wav/3_9.wav/4_0.wav/4_10.wav/4_11.wav/4_13.wav/4_2.wav/4_3.wav/4_4.wav/4_5.wav/4_6.wav/4_7.wav/4_8.wav/4_9.wav/5_0.wav/5_1.wav/5_10.wav/5_11.wav/5_12.wav/5_13.wav/5_14.wav/5_15.wav/5_16.wav/5_17.wav/5_18.wav/5_19.wav/5_2.wav/5_20.wav/5_21.wav/5_22.wav/5_23.wav/5_24.wav/5_25.wav/5_26.wav/5_27.wav/5_28.wav/5_29.wav/5_3.wav/5_30.wav/5_31.wav/5_32.wav/5_33.wav/5_34.wav/5_35.wav/5_36.wav/5_37.wav/5_38.wav/5_39.wav/5_4.wav/5_40.wav/5_41.wav/5_42.wav/5_43.wav/5_44.wav/5_45.wav/5_46.wav/5_47.wav/5_48.wav/5_49.wav/5_5.wav/5_50.wav/5_51.wav/5_52.wav/5_53.wav/5_54.wav/5_55.wav/5_56.wav/5_57.wav/5_58.wav/5_6.wav/5_60.wav/5_7.wav/5_8.wav/5_9.wav/6_0.wav/6_1.wav/6_10.wav/6_11.wav/6_12.wav/6_13.wav/6_14.wav/6_15.wav/6_16.wav/6_17.wav/6_18.wav/6_19.wav/6_2.wav/6_20.wav/6_21.wav/6_22.wav/6_23.wav/6_24.wav/6_25.wav/6_26.wav/6_27.wav/6_28.wav/6_29.wav/6_3.wav/6_30.wav/6_31.wav/6_32.wav/6_33.wav/6_34.wav/6_35.wav/6_36.wav/6_37.wav/6_38.wav/6_39.wav/6_4.wav/6_40.wav/6_41.wav/6_42.wav/6_43.wav/6_44.wav/6_45.wav/6_47.wav/6_5.wav/6_6.wav/6_7.wav/6_8.wav/6_9.wav/7_0.wav/7_1.wav/7_10.wav/7_100.wav/7_102.wav/7_103.wav/7_104.wav/7_105.wav/7_106.wav/7_108.wav/7_11.wav/7_12.wav/7_13.wav/7_14.wav/7_15.wav/7_16.wav/7_17.wav/7_18.wav/7_19.wav/7_20.wav/7_21.wav/7_22.wav/7_23.wav/7_24.wav/7_25.wav/7_26.wav/7_27.wav/7_28.wav/7_29.wav/7_3.wav/7_30.wav/7_31.wav/7_32.wav/7_33.wav/7_34.wav/7_35.wav/7_36.wav/7_37.wav/7_38.wav/7_39.wav/7_40.wav/7_41.wav/7_43.wav/7_44.wav/7_45.wav/7_46.wav/7_47.wav/7_48.wav/7_49.wav/7_5.wav/7_50.wav/7_51.wav/7_52.wav/7_53.wav/7_54.wav/7_55.wav/7_56.wav/7_57.wav/7_58.wav/7_59.wav/7_6.wav/7_60.wav/7_61.wav/7_62.wav/7_63.wav/7_64.wav/7_65.wav/7_66.wav/7_67.wav/7_68.wav/7_69.wav/7_7.wav/7_70.wav
```
Is one of them :) (this is the list that should contain input paths for each audio)

I'm working on it though

UPD: fixed in 8e32601

413x1nkp committed:
style(TODO): add TODO semantics for some future fixes comments

413x1nkp committed:
fix: fix feature input list bug

still doesn't save npy files though...

413x1nkp committed:
fix: pass the full filename with .npy extension

this doesn't fix the issue of features not saving to disk, but it should help in the long run

413x1nkp commented:
UPDATE: still having troubles with saving the f0 to the folder... Not sure why that is, it doesn't appear like anything is broken? a509cfb did not fix that either...
Any ideas, @fumiama ?

fumiama commented:
You changed it too much, so even me cannot compare the code line by line 😂. As for saving, you can just copy the naming logic of the original code to make sure your new name is equal to the original one. Then if you do not like its code style, do a equalized refactor, which means that you make sure the result is the same after each changes. If the result changed, just revert this step and see why, then fix it.

413x1nkp commented:
> You changed it too much, so even me cannot compare the code line by line 😂. As for saving, you can just copy the naming logic of the original code to make sure your new name is equal to the original one. Then if you do not like its code style, do a equalized refactor, which means that you make sure the result is the same after each changes. If the result changed, just revert this step and see why, then fix it.

Alrighty, I'll work on it
Thanks :)

413x1nkp committed:
feat: merge rmvpe and rmvpe_gpu into single f0 method

413x1nkp commented:
Turns out the issue is the creation of `torch.multiprocessing.Process()`. Calling `save_f0()` like this works:
```py
    for idx, (inp_path, coarse_path, feature_path) in enumerate(paths):
        if idx % n == 0:
            __log(
                log_file, f"Computing f0; Current: {idx}; Total: {n_paths}; {inp_path}"
            )

        save_f0(predictor, inp_path, coarse_path, feature_path, f"{expected_dir}/extract_f0_feature.log")
        logger.debug(f"Starting extract_f0_feature_{idx} thread for {inp_path}")
```

Not sure if there's a better fit for I/O operations with multiprocessing in mind...

413x1nkp committed:
fix: remove parallel f0 saving

this is done due to the poor design i've imposed on myself and misunderstanding of the code i wrote myself. now f0 extraction works, even if it's a bit slow

GARBAGE CODE, REWRITE LATER!!

413x1nkp committed:
fix: remove sys.TextIOWrapper type specifier

413x1nkp commented:
Figured out why RMVPE isn't working:
```
Failed to compute f0 for - /kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/logs/mi-test/1_16k_wavs/8_1.wav: '<=' not supported between instances of 'float' and 'NoneType'
```
And this error is created for each file. Not sure why this is `rmvpe`-specific, but I'll look into it

413x1nkp committed:
refactor: remove rmvpe_gpu from predictors dict since it was merged into rmvpe

fumiama commented:
> Not sure if there's a better fit for I/O operations with multiprocessing in mind...

How about using a `multiprocessing.Queue()` and collect the result in main process, then save them?

fumiama commented:
> And this error is created for each file. Not sure why this is rmvpe-specific, but I'll look into it

Have you defined the `filter_radius` value to `0.03`? It is None by default. But well, I will write a fallback about that.

413x1nkp committed:
Merge branch 'dev' into dev

fumiama commented:
As for the logger, you can refer to

https://github.com/fumiama/Retrieval-based-Voice-Conversion-WebUI/blob/412a9950a1e371a018c381d1bfb8579c4b0de329/infer/lib/train/utils.py#L418-L430

The handler can be multiple so we can print it to stdout while writing it to file.

413x1nkp committed:
Merge branch 'dev' into dev

413x1nkp commented:
> As for the logger, you can refer to
> 
> https://github.com/fumiama/Retrieval-based-Voice-Conversion-WebUI/blob/412a9950a1e371a018c381d1bfb8579c4b0de329/infer/lib/train/utils.py#L418-L430
> 
> The handler can be multiple so we can print it to stdout while writing it to file.

I see that it takes in the model dir as an argument too. I assume You want me to make a new function just like this but that will just print strings of text like "F0-todo: <number_of_f0>" and such?

fumiama commented:
> > As for the logger, you can refer to
> 
> > 
> 
> > https://github.com/fumiama/Retrieval-based-Voice-Conversion-WebUI/blob/412a9950a1e371a018c381d1bfb8579c4b0de329/infer/lib/train/utils.py#L418-L430
> 
> > 
> 
> > The handler can be multiple so we can print it to stdout while writing it to file.
> 
> 
> 
> I see that it takes in the model dir as an argument too. I assume You want me to make a new function just like this but that will just print strings of text like "F0-todo: <number_of_f0>" and such?

Well, the format doesn't matter. Just use the handler to write file but not write it manually.

fumiama reviewed (commented):
I have done the same thing as what your PR aiming to do so I would like to close this one. However, still thanks for your effort!
